contracts/protocore/SelfProtocore.sol

```diff
@@ -69,5 +69,43 @@ contract SelfProtocore is MasterCopyNonUpgradable, GenesisSelfProtocore, Validat
             genesisAuxiliaryTargetBlockHash,
             genesisAuxiliaryTargetBlockNumber
         );
+        ValidatorSet.setupValidatorSet();
     }
+
+
+    /* External */
+
+    /**
+     * @notice Insert or remove validator. It inserts validator if not already
+     *         present and reputation is greater than 0. It removes validator
+     *         if it is present and reputation is 0.
+     *
+     * @dev Function requires:
+     *      - Caller should be Coconsensus contract.
+     *      - Validator must enter at height equal to current open kernel height.
+     *
+     * @param _validator Validator address to upsert.
+     * @param _height Validator start or end height to be updated.
+     * @param _reputation Validator's reputation value.
+     */
+    function upsertValidator(
+        address _validator,
+        uint256 _height,
+        uint256 _reputation
+    )
+        external
+        onlyCoconsensus
+    {
+        if(ValidatorSet.inValidatorSet(_validator, _height)) {
```

Comment:
Somehow these conditions are bothering me.
It should check if the validator set already existed irrespective of height provided.
Similarly, it should check if the end height is MAX_FUTURE_END_HEIGHT in order to remove the validator.

Reply:
Thank you.
Added check to verify if validator existed irrespective of it's height.
Regarding check to verify end heigh of validator is MAX_FUTURE_END_HEIGHT, it's already happening in `ValidatorSet::removeValidatorInternal`.
https://github.com/mosaicdao/mosaic-1/blob/develop/contracts/validator/ValidatorSet.sol#L157

Reply:
I was suggesting this because I wanted to make sure that if a validator is removed once, then calling it again will not revert. Our discussion was that Upsert should not revert. Maybe this will not happen. I will need a second eye to check this logic. cc: @pgev